File: backend/repositories/stock_repository.py
from typing import List, Dict, Optional
import pyodbc
from repositories.base_repository import BaseRepository, DatabaseConnectionError, QueryError
from models.stock import StockItem, StockSearchRequest


class StockRepository(BaseRepository):
    """Repository for stock data queries."""

    def search_by_activity(self, request: StockSearchRequest) -> List[Dict]:
        """
        Search stock by article code, location, or storage number within activity.

        Returns list of dicts with keys: art_code, stk_lieu, stk_nosu, qua_code, stk_qte

        Raises:
            QueryError: on database query failure
            DatabaseConnectionError: on connection failure
        """
        # Validation is handled by Pydantic model_validator

        # Build WHERE clause with parameterized query (activity filter is AND, rest are OR)
        conditions = ["ACT_CODE = ?"]
        params = [request.act_code]

        if request.art_code:
            conditions.append("ART_CODE LIKE ?")
            params.append(f"%{request.art_code}%")
        if request.stk_lieu:
            conditions.append("STK_LIEU LIKE ?")
            params.append(f"%{request.stk_lieu}%")
        if request.stk_nosu:
            conditions.append("STK_NOSU LIKE ?")
            params.append(f"%{request.stk_nosu}%")

        # First condition is activity filter (AND), rest are OR
        where_clause = conditions[0] + " AND (" + " OR ".join(conditions[1:]) + ")"

        query = f"""
            SELECT ART_CODE, STK_LIEU, STK_NOSU, QUA_CODE, STK_QTE
            FROM STK_DAT
            WHERE {where_clause}
            ORDER BY ART_CODE, STK_LIEU
        """

        try:
            rows = self.execute_query(query, tuple(params))
        except (DatabaseConnectionError, QueryError) as exc:
            raise QueryError(f"Stock search failed: {str(exc)}") from exc

        # Convert rows to dicts
        results = []
        for row in rows:
            results.append({
                "art_code": row[0],
                "stk_lieu": row[1],
                "stk_nosu": row[2],
                "qua_code": row[3],
                "stk_qte": row[4],
            })

        return results

File: backend-go/repositories/stock_repository.go
package repositories

import (
	"database/sql"
	"fmt"
	"strconv"
	"strings"

	"stock-api/models"
)

// StockRow is a single stock entry returned by the stock repository.
type StockRow struct {
	ArtCode string
	StkLieu string
	StkNosu string
	QuaCode string
	StkQte  int
}

// StockRepository provides access to stock data (table STK_DAT).
type StockRepository struct {
	DB *sql.DB
}

// NewStockRepository wires the shared connection pool into the repository.
func NewStockRepository(db *sql.DB) *StockRepository {
	return &StockRepository{DB: db}
}

// SearchByActivity searches stock by article code, location and/or storage
// number within an activity. The activity filter is mandatory (AND); the other
// criteria are combined with OR. All values are passed as parameters using the
// go-mssqldb @pN positional placeholder style.
func (r *StockRepository) SearchByActivity(req models.StockSearchRequest) ([]StockRow, error) {
	// First parameter is always the activity filter (AND).
	params := []interface{}{req.ActCode}

	// Collect the optional OR criteria.
	var orCriteria []string
	addCriterion := func(column string, value *string) {
		if value == nil || *value == "" {
			return
		}
		// Next placeholder index is len(params)+1 (1-based).
		placeholder := "@p" + strconv.Itoa(len(params)+1)
		orCriteria = append(orCriteria, column+" LIKE "+placeholder)
		params = append(params, "%"+*value+"%")
	}

	addCriterion("ART_CODE", req.ArtCode)
	addCriterion("STK_LIEU", req.StkLieu)
	addCriterion("STK_NOSU", req.StkNosu)

	var query strings.Builder
	query.WriteString("SELECT ART_CODE, ISNULL(STK_LIEU,''), ISNULL(STK_NOSU,''), ISNULL(QUA_CODE,''), CAST(STK_QTE AS INT) ")
	query.WriteString("FROM STK_DAT ")
	query.WriteString("WHERE ACT_CODE = @p1")
	if len(orCriteria) > 0 {
		query.WriteString(" AND (")
		query.WriteString(strings.Join(orCriteria, " OR "))
		query.WriteString(")")
	}
	query.WriteString(" ORDER BY ART_CODE, STK_LIEU")

	rows, err := r.DB.Query(query.String(), params...)
	if err != nil {
		return nil, fmt.Errorf("stock search failed: %w", err)
	}
	defer rows.Close()

	var results []StockRow
	for rows.Next() {
		var item StockRow
		if err := rows.Scan(&item.ArtCode, &item.StkLieu, &item.StkNosu, &item.QuaCode, &item.StkQte); err != nil {
			return nil, fmt.Errorf("stock search scan failed: %w", err)
		}
		results = append(results, item)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("stock search iteration failed: %w", err)
	}

	return results, nil
}
